package-res/resources/web/pentaho/type/_type.js

```diff
@@ -0,0 +1,1360 @@
+       * serialization-only id is generated.
+       * In the first occurrence in the given scope,
+       * that id is returned, within a full specification of the type,
+       * obtained by calling [toSpecInContext]{@link pentaho.type.Type#toSpecInContext}.
+       * In following occurrences, only the previously used temporary id is returned.
+       *
+       * Some standard types have a special reference syntax,
+       * e.g. [List.Type#toRef]{@link pentaho.type.List.Type#toRef}.
+       *
+       * @see pentaho.type.Type#toSpec
+       *
+       * @param {Object} [keyArgs] - The keyword arguments object.
+       * Passed to every type and instance serialized within this scope.
+       *
+       * Please see the documentation of subclasses for information on additional, supported keyword arguments.
+       *
+       * @return {!pentaho.type.spec.UTypeReference} A reference to this type.
+       */
+      toRef: function(keyArgs) {
+        return this.shortId || O.using(new SpecificationScope(), this.toRefInContext.bind(this, keyArgs || {}));
+      },
+
+      /**
+       * Returns a _reference_ to this type under a given specification context.
+       *
+       * This method requires that there currently exists an
+       * [ambient specification context]{@link pentaho.type.SpecificationContext.current}.
+       *
+       * @see pentaho.type.Type#toRef
+       *
+       * @param {Object} [keyArgs] - The keyword arguments object.
+       * Passed to every type and instance serialized within this scope.
+       *
+       * Please see the documentation of subclasses for information on additional, supported keyword arguments.
+       *
+       * @return {!pentaho.type.spec.UTypeReference} A reference to this type.
+       */
+      toRefInContext: function(keyArgs) {
+        return this.shortId || SpecificationContext.current.getIdOf(this) || this.toSpecInContext(keyArgs);
+      },
+      //endregion
+
+      /**
+       * Returns a textual representation suitable to identify this type in an error message.
+       *
+       * @return {string} A textual representation.
+       */
+      toString: function() {
+        return this.id || this.label; // Never empty;
+      }
+    }, /** @lends pentaho.type.Type */{
+
+      //@override
+      /**
+       * See Base.js
+       * @ignore
+       */
+      _subclassed: function(SubTypeCtor, instSpec, classSpec, keyArgs) {
+
+        O.setConst(this.prototype, "_hasDescendants", true);
+
+        var SubInstCtor = keyArgs.instance.constructor;
+
+        // Links SubTypeCtor and SubInstCtor and "implements" instSpec.
+        SubTypeCtor._initInstCtor(SubInstCtor, instSpec, keyArgs);
+
+        // Implement the given static interface.
+        if(classSpec) SubTypeCtor.implementStatic(classSpec);
+      },
+
+      // Links TypeCtor (this) and the given InstCtor together and
+      // applies the TypeCtor constructor to its own prototype, as a way to initialize it.
```

Comment:
Suggest:

applies the TypeCtor constructor to its own prototype as a way to initialize it.